pkg/session_manager/session_manager.go

```diff
@@ -0,0 +1,126 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package session_manager
+
+import (
+	"sync"
+
+	"go.uber.org/zap"
+)
+
+type (
+	Func          func(string) (interface{}, error)
+	EventCallback func(key string, bundlesChannel <-chan interface{})
+)
+
+// command represents command used to communicate within the session manager
+type command struct {
+	op    string
+	key   string
+	value interface{}
+}
+
+// SessionManager is a manager for session channels
+// It allows to subscribe, publish and unsubscribe to/from channels
+type SessionManager struct {
+	eventCallback EventCallback
+	mu            sync.Mutex
+	channels      map[string]chan interface{}
+	logger        *zap.Logger
+	operations    chan command
+}
+
+// New creates a new SessionManager
+func New(
+	logger *zap.Logger,
+	eventsCallback EventCallback,
+) *SessionManager {
+	manager := &SessionManager{
+		eventCallback: eventsCallback,
+		mu:            sync.Mutex{},
+		channels:      make(map[string]chan interface{}),
+		logger:        logger,
+		operations:    make(chan command),
+	}
+
+	go manager.processCommands()
+
+	return manager
+}
+
+// Sub subscribes to a channel key
+// It's fine to call this function multiple times with the same key - only one
+// channel will be created
+func (manager *SessionManager) Sub(key string) {
+	manager.operations <- command{op: "sub", key: key}
+}
+
+func (manager *SessionManager) sub(key string) {
+	_, found := manager.channels[key]
+	if !found {
+		manager.logger.Debug("Subscribing to key", zap.String("key", key))
+		ch := make(chan interface{})
+		manager.channels[key] = ch
+		go manager.eventCallback(key, ch)
+	}
+}
+
+// Pub publishes a value to a channel key
+func (manager *SessionManager) Pub(key string, value interface{}) {
+	manager.operations <- command{op: "pub", key: key, value: value}
+}
+
+func (manager *SessionManager) pub(key string, value interface{}) {
+	ch, found := manager.channels[key]
+	if found {
+		ch <- value
+	} else {
+		manager.logger.Warn("Channel for publishing does not exist", zap.String("key", key))
+	}
+}
+
+// Unsub unsubscribes from a channel key
+// It's fine to call this function multiple times with the same key - only if the key
+// still exists, it will be closed.
+func (manager *SessionManager) Unsub(key string) {
+	manager.operations <- command{op: "unsub", key: key}
+}
+
+func (manager *SessionManager) unsub(key string) {
+	ch, found := manager.channels[key]
+	if found {
+		manager.logger.Debug("Unsubscribing to key", zap.String("key", key))
+		delete(manager.channels, key)
+		close(ch)
+	}
+}
+
+func (manager *SessionManager) processCommands() {
+	for {
+		cmd := <-manager.operations
+		// manager.logger.Debug("SessionManager - processCommands - START", zap.String("cmd", cmd.op), zap.String("key", cmd.key))
+		switch cmd.op {
```

Comment:
Using `channelOpen` will be only growing => it will be consuming more and more memory => this is not something, that we want.

Now there are two ways how you can figure out, that you are publishing messages to closed channel:
1. There will be these warnings.
2. There will be lot of unprocessed events (marked as `waiting`), so you will get error at the end.

Reply:
Writing to the closed chanel causes panic and we should avoid this.